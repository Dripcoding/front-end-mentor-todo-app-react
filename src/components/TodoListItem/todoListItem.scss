@import "../../mixins";
@import "../../palette.scss";

.todoListItem__container {
  @include flex(column, center, flex-start);
  background-color: $very-dark-desaturated-blue;
  height: 64px;
  width: 540px;
  padding: 20px 312px 20px 24px;

  .todoListItem__input {
    @include flex(row, flex-start, center);
    height: 24px;
    width: 492px;

    input {
      @include text($light-input-color, 18px, normal, 400, -0.25px, 18px, left, none);
      background-color: inherit;
      border: none;
    }
  }
}