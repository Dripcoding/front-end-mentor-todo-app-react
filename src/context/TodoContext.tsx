import React, { useReducer } from 'react';
import { todoReducer } from './reducer';

interface ITodoContextProps {
    children: React.ReactChild
}

export interface ITodo {
    id: number;
    text: string;
    active: boolean;
}

export interface ITodos {
    todos: ITodo[];
}

const initialState: ITodos = {
    todos: [{id: 1, text: "finish homework", active: true}]
}

const TodoContext: React.Context<ITodos> = React.createContext(initialState);

const TodoContextProvider = ({children}: ITodoContextProps): JSX.Element => {
    const [state, dispatch] = useReducer(todoReducer, initialState);

    return (
        <TodoContext.Provider value={initialState}>
            {children}
        </TodoContext.Provider>
    )
}


export default TodoContextProvider;