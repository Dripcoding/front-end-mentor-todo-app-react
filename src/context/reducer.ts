import { ITodo, ITodos } from './TodoContext';

// todo action types
export const CREATE_TODO: string = 'CREATE_TODO';
export const EDIT_TODO: string = 'EDIT_TODO';
export const DELETE_TODO: string = 'DELETE_TODO';
export const COMPLETE_TODO: string = 'COMPLETE_TODO';
export const ACTIVATE_TODO: string = 'ACTIVATE_TODO';
export const DELETE_ALL_TODOS: string = 'DELETE_ALL_TODOS';
// visibility action types
export const VIEW_ALL_TODOS: string = 'VIEW_ALL_TODOS';
export const VIEW_ALL_COMPLETED_TODOS: string = 'VIEW_ALL_COMPLETED_TODOS';
export const VIEW_ALL_ACTIVE_TODOS: string = 'VIEW_ALL_ACTIVE_TODOS';


interface IAction {
    type: string;
    payload: ITodo 
}

export const todoReducer = (state: ITodos, action: IAction): ITodos => {
    switch(action.type) {
        case CREATE_TODO: {
            return { todos: [...state.todos, action.payload] }     
        }
        case EDIT_TODO: {
            const updatedTodo = action.payload;
            const todos = state.todos.map(todo => {
                if (todo.id === updatedTodo.id) {
                    todo.text = updatedTodo.text;
                }
                return todo;
            })
            return { todos }
        }
        case DELETE_TODO: {
            const todos = state.todos.filter(todo => {
                return todo.id !== action.payload.id
            })
            return { todos }
        }
        case COMPLETE_TODO: {
            const todos = state.todos.map(todo => {
                if (todo.id === action.payload.id) {
                    todo.active = false
                }
                return todo;
            })
            return { todos }
        }
        default:
            return state;
    }
}
