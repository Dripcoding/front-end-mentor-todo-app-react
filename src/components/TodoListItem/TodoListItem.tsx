import React, { useContext, useState } from "react";
import { ITodo, TodoContext } from "../../context/TodoContext";

import "./todoListItem.scss";
import TodoCheckBoxToggle from "../TodoCheckBoxToggle";
import { EDIT_TODO } from "../../context/reducer";

interface ITodoListItemProps {
  todo: ITodo;
}

const STYLE_BASE: string = "todoListItem__";

const TodoListItem = ({ todo }: ITodoListItemProps): JSX.Element => {
  const { dispatch } = useContext(TodoContext);
  const [value, setValue] = useState<string>(todo.text);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setValue(e.target.value);
  };

  const editTodo = (e: React.KeyboardEvent): void => {
    if (e.key === "Enter") {
      dispatch({ type: EDIT_TODO, payload: { id: 1, text: value, active: true } });
    }
  };

  return (
    <div className={`${STYLE_BASE}container`} data-testid={"TODO_LIST_ITEM"}>
      <div className={`${STYLE_BASE}input`}>
        <TodoCheckBoxToggle disabled={false} />
        <input
          type="text"
          className={`${STYLE_BASE}input`}
          data-testid={"TODO_LIST_ITEM"}
          onChange={handleChange}
          onKeyDown={editTodo}
          value={value}
        />
      </div>
    </div>
  );
};

export default TodoListItem;