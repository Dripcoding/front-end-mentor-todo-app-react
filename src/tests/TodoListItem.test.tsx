import { render } from "@testing-library/react";
import TodoListItem from "../components/TodoListItem";
import { ITodo } from "../context/TodoContext";
import userEvent from "@testing-library/user-event";

const todo: ITodo = {
  id: 10,
  text: "example",
  active: false
}

describe("TodoListItem", () => {
  it("matches snapshot", () => {
    const { container } = render(<TodoListItem todo={todo} />);
    expect(container).toMatchSnapshot();
  });

  it("renders correctly", () => {
    const { getByTestId, getByRole } = render(<TodoListItem todo={todo} />);

    const container = getByTestId("TODO_LIST_ITEM");
    expect(container).toBeInTheDocument();
    expect(container).toBeVisible();

    const todoToggle = getByTestId("TODO_CHECKBOX_TOGGLE");
    expect(todoToggle).toBeInTheDocument();
    expect(todoToggle).toBeVisible();

    const todoListItemInput = getByRole("textbox");
    expect(todoListItemInput).toBeInTheDocument();
    expect(todoListItemInput).toBeVisible();
    expect(todoListItemInput).toHaveValue("example");
  });

  it("captures todo text correctly", () => {
    const { getByRole } = render(<TodoListItem todo={todo} />);

    const sampleInput = "finish homework";
    const todoListItemInput = getByRole("textbox");
    userEvent.clear(todoListItemInput);
    userEvent.type(todoListItemInput, sampleInput);

    expect(todoListItemInput).toHaveValue(sampleInput);
  });

  it("edits todo text correctly", () => {
    const { getByDisplayValue } = render(<TodoListItem todo={todo} />);

    const sampleInput = "edited todo";
    const todoListItemInput = getByDisplayValue(todo.text);

    expect(todoListItemInput).toBeInTheDocument();

    userEvent.clear(todoListItemInput);
    userEvent.type(todoListItemInput, sampleInput + "{enter}");

    expect(todoListItemInput).toHaveValue(sampleInput);
  });
});