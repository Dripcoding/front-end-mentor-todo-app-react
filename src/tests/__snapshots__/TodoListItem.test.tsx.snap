// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem matches snapshot 1`] = `
<div>
  <div
    class="todoListItem__container"
    data-testid="TODO_LIST_ITEM_CONTAINER"
  >
    <div
      class="todoListItem__input"
    >
      <div
        class="todoCheckBoxToggle__container"
        data-testid="TODO_CHECKBOX_TOGGLE"
      />
      <input
        class="todoListItem__input"
        data-testid="TODO_LIST_ITEM"
        type="text"
        value="example"
      />
    </div>
  </div>
</div>
`;
